### DISCUSSION TITLE
TCC-Link AB : Is it a RS485 Bus?

### DISCUSSION THREAD

**Discussion by Inscribe** [issalig/toshiba_air_cond#5] [Q&A] [answered]
Hi Issalig,

I appreciate all the hard work put in to get this interface going. Please keep it up!

In many Toshiba literature, I see mention of RS485/422 for the TCC-Link (U1, U2), (U3,U4) and (A,B). The cables called for are also un-polarized. 

My question is : Have you confirmed that the A and B are, indeed, RS485 and can be handled by a transceiver like the   SN75HVD12 ? 

I do like your method of using opto-isolators for converting A & B to TTL3.3 and your intent to derive supply from the A & B lines initially. But, a RS485 transceiver would help in reducing the part count.

Thanks

- Mish

**Comment by issalig** [comment]
Hi, glad to see you like my work. When I started this I had no information about TCC and I had to discover everything. If you have some additional documentation about TCC-Link it would be nice to include it in the project.

In the beginning I thought about RS485 (but honestly I do not know the features, just that it is an industrial long range  rs232). Then I plugged in the oscilloscope and I saw a kind of serial signal over 13v so I just invented a circuit to read it and it happily worked.

Using SN75HVD12 would reduce the parts but also the entertainment when soldering :) I will consider it for the future iterations. Also I had some RS485 modules laying around and I will give them a try.

My next goal is to power the ESP from A,B that actually does not work, but as I am not an electrical engineer I still have to learn what's the reason. Maybe ESP8266 is drawing too much current (around 200-300mA). If you have any idea, it will be more than welcome.

BR.

**Reply by Inscribe** [reply]
Thanks for your reply.

Since there is no ground connection at the terminals, I assume only A and B are available as differential signals. Off the hat, I think it would not be possible to derive a DC voltage from these A/B. I shall continue working on this and revert.

Using an oscilloscope is difficult on the differential signals. See https://youtu.be/yNtuVDKkOnk for one method. Then, you can try your RS485 module to get RX and TX access.

**Reply by Arnold-n** [reply]
> If you have some additional documentation about TCC-Link it would be nice to include it in the project.

Great work! Actually I think that the A/B protocol you reverse engineered is different from the U1/U2 and U3/U4 protocol which is based on HBS (Japanese Home Bus System). I wonder if the A/B interface is referred to as TCC-link at all?

Anyway, TCC-link (on U1/U2 and U3/U4) data can be monitored with [M-NET-Sniffer](https://github.com/LenShustek/M-NET-Sniffer) or with my own project [P1P2Serial](https://github.com/Arnold-n/P1P2Serial), the latter also supports writing packet data to the TCC-link bus (if you figure out how the protocol works).

**Reply by issalig** [reply]
I remember that I was checking the HBS and it had some similarities but it was not the same thing.

There is a guy that started decoding communication btw central unit and external unit but he did not publish the information. You have some info in this issue https://github.com/issalig/toshiba_air_cond/issues/10

**Reply by Arnold-n** [reply]
I think what he reverse engineered is a one-wire and 230V-connected serial one-wire bidirectional point-to-point connection for the single-unit wall-airco. I see no reference there to TCC-link or to U1/U2 or U3/U4. [This PCB](https://www.toshiba-aircon.co.uk/product/tcb-pcnt30tle-tcc-link-network-adaptor/) shows a MM1192 HBS chip and the silk screen mentions HBS) so the TCC-link is really HBS - and there is one Toshiba VRF system where we indeed obtained data from U1/U2 with a MM1192 chip. The logical format is likely different - for example the last packet byte is not a zero-bit-count but is a byte-wise XOR.

**Reply by issalig** [reply]
That's nice info. I did not know about MM1192, I will take a look at it.

**Reply by Arnold-n** [reply]
There is also the MAX22088 (which I now use in my design) but the MM1192 (DIP/SOIC) is more breadboard friendly and the MAX22088 (QFN) is only really useful for bus-powered HBS (some Toshiba TCC-links provide bus-power, others apparently not).

**Comment by issalig** [comment, marked as answer]
Power is obtained from AB because remote control unit RBC-AMT32E is being powered from there, in fact, I have connected  RBC-AMT32E and my module in parallel to AB line. 

Plugging oscilloscope into AB let you see the bits without problem. Just with a 13V offset

I have this RS485 module, I will take it a look but I am not sure what to connect to GND next to AB ![](https://www.makerlab-electronics.com/wp-content/uploads/2017/11/TTL-RS485-Module-4-768x370.jpg)

**Reply by cmc88** [reply]
Hi @issalig,
first of all well done for your project. Was a long time looking for a project to make my toshiba RAV-SM1406BTP-E a bit more smart. My question is regarding the board above. Does the board shown in the picture above works with the unit?

Thanks

**Reply by issalig** [reply]
I do not think so but I do not know. I built my own circuit and it worked so I did not try other boards.

If you try please please tell.

**Reply by cmc88** [reply]
Will it be please possible to buy one of your boards?

**Reply by issalig** [reply]
Yes, you can contact me at cedux AT zetmail   com

**Reply by sirajshikhli** [reply]
@issalig @muxa 

I have an urgent project and well paid for TCC design for IoT platform, I really need your help to be part of our team 

Please email me at: sirajshikhli@mastersshields.com

**Reply by MrBlue90** [reply]
> Yes, you can contact me at cedux AT zetmail com

Hi @issalig,
This email address doesn't seem to be active anymore.
Do you have a different address to get in touch with you?
Thanks

**Comment by Inscribe** [comment]
Yes. This struck me after I posted the message. They are using two features beyond 485 : polarity independence and power over 485. 
Where are you located? What are your resources to solder a breadboard? I don’t have access to the AC. My new house has this brand and they will give me possession in two months. 
Action: Can you reconfirm the oscilloscope readings as per the YouTube link?

**Comment by issalig** [comment]
I am located in Europe and I can solder PCBs.

I have not checked Youtube, just plugged my small oscilloscope aliexpress.com/item/32878694776.html and could see the signal.

**Reply by Inscribe** [reply]
Hi Issalig,
Need a few inputs from you:
A. How did you probe A and B? Where were the pin and GND of the probe connected?
B. Are you sure that the baud rate you are using on ESP for read/write is 2400 baud?
C. Can you send a picture of the Wired Remote board, showing the components? Will help estimate the power extraction used by the device.

**Reply by issalig** [reply]
A. Signal to A and GND to B
B. For sure, I am using it.
C. You can find the Kicad files here https://github.com/issalig/toshiba_air_cond/tree/master/pcb/airpcb_smd  and the schematic https://github.com/issalig/toshiba_air_cond/blob/master/pcb/airpcb_smd/schematics.pdf

**Reply by Inscribe** [reply]
Thanks, Issalig. Got A and B right. At C above, wanted to see the Toshiba wired remote controller pcb, if you have a picture or two. You had mentioned that your esp is taking power from the wired remote. That is why this question. In particular, there should be two inductors on the pcb. Would be interesting to have those part numbers or values or just the size. 
Action: Please post some pictures of the wired controller pcb.
P.S: I will be sending you two documents for understanding power over 485. Then, a bit later, a schematic which we can try out.

**Reply by Inscribe** [reply]
[tidu993.pdf](https://github.com/issalig/toshiba_air_cond/files/6781667/tidu993.pdf)

**Reply by Inscribe** [reply]
https://hackaday.io/project/27176-digitize-your-analog-marklin-ho-model-trains/log/67046-dc-supply-over-rs485

**Reply by issalig** [reply]
Thanks for the info.

I have uploaded remote pcb photo and I do not see any inductor on it.
Also there is a photo of my connection to the remote.

https://github.com/issalig/toshiba_air_cond/blob/master/pcb/mounted_board.jpg
https://github.com/issalig/toshiba_air_cond/blob/master/pcb/remote_back_pcb.jpg

**Reply by issalig** [reply]
BTW IC2 is a 7805 and I tried to power the ESP from there but ESP does not turns on. Maybe too little current?

**Comment by muxa** [comment]
I was going to build the original circuit as designed by @issalig, but then stumbled upon this discussion. 
Has anybody tried using SN75HVD12 on TCC link?  Curious to hear your findings.

**Reply by issalig** [reply]
I do not really know, as it works with my circuit I am not very interested in other solutions.

**Reply by Inscribe** [reply]
Hi Muxa,
As Issalig said, his circuit works and so does his software. You may go ahead and implement the same and it should work too.
My question was more investigative, so that (a) standard RS485 transceiver can be used and (b) Power can be derived from the bus. 
Unfortunately, I still don’t have access to the AC. It will possibly take a few months before I get to that. Will keep you all informed.

**Reply by issalig** [reply]
Hope you get it soon and can play with it and give feedback to the project.

My TODO list is the following:
- Power from AB line instead of mobile phone charger. (maybe it is not possible due to hungry ESP8266)
- Polish web interface
- Fix random crashes
- HomeAssistant support

**Comment by Inscribe** [comment]
OK, I have found some new information today and would answer my own question raised.

No, the TCC-Link IS NOT a RS485 equivalent two-wire network. More details will follow.

**Reply by Inscribe** [reply]
Issalig, PM possible? See my profile. Thanks.

**Reply by issalig** [reply]
Hi, you can use this thread so everybody can follow the discussion and participate.

**Reply by issalig** [reply]
Inscribe, I received your mail, but for some reason I get an error when answering so I answer here.
Nice you found a candidate ic, I have read the spec sheet and it looks good.

**Comment by muxa** [comment]
@issalig 
> Power from AB line instead of mobile phone charger. (maybe it is not possible due to hungry ESP8266)

I managed to successfully power ESP8266 from the AB live. Note: I only have a reader circuit. The writer circuit is still to be built, but I don't think it's going to change power consumption significantly, so should be all to powering the entire module from AB line. 

I used a Mini-360 buck converter first, but for some reason the RX input pin had too much noise making the data unreadable (I used GPIO13 ping with PULLUP for RX). I then replaced it with [HeSai buck converter](https://www.aliexpress.com/item/32998937888.html) and the line became clean. There are occasional voltage spikes on the RX line that cause UART to received some bogus data (usually `0xFF`), but it's easy to filter it out in software. 

Here's the reader circuit that I'm using:
<img width="518" alt="Screen Shot 2022-08-13 at 12 29 23 PM" src="https://user-images.githubusercontent.com/128341/184459993-775839f3-38ea-4906-9d2e-dfafb711d4c0.png">

UPDATE 2022-08-16:

Here's the full schematics:

[Schematic_Toshiba AC Control_2022-08-16.pdf](https://github.com/issalig/toshiba_air_cond/files/9339513/Schematic_Toshiba.AC.Control_2022-08-16.pdf)

**Comment by issalig** [comment]
I have also powered it from AB line and while it works for TX, I have problems for RX, probably noise like you
I have used this circuit that was provided by @bbqkees
In order to save

![image](https://user-images.githubusercontent.com/7136948/184478133-95a76836-ec0d-452a-bf1d-3ded4d1e4fb6.png)

I will check my RX circuit and I need to put the oscilloscope and see what happens.

I have used Mini-360 in the past and I would avoid them, there are cheap and also a crap. Take care because converting 14V down to 3.3V it is a big step and they will get very hot.

Also I have used a 3.3V regulator that goes to the 3.3V line of the ESP8266, thus bypassing the 5v->3.3V regulator from the esp8266 that is known to be very power hungry.

**Reply by mateuszdrab** [reply]
Have you got an updates on this? 
I am just looking for a solution to extract power from the bus to provide 12v to a Modbus-Wifi relay as I have the AB line connected to a Realtime FDP3 Modbus controller for the AC.

**Comment by smulikHakipod** [comment]
I tested this RS485 module https://www.aliexpress.com/item/1005006656455603.html?spm=a2g0o.order_list.order_list_main.64.419b7385uglyHS and it did not work. 

The controller was shutting down. It seems there is too low residence of around 120ohm between AB wires in the modules, which is probably too low, causing everything to shutdown

**Reply by issalig** [reply]
I have read that RS485 works with voltages -7V to 12V.  Could that be the problem because the signal is > 12V?
